README.rst

```diff
@@ -471,6 +471,120 @@ any application code:
+
+Additionally, if we wanted to have the :code:`native1.r` rule code print to stdout instead, we could also set parameters for the rule and change
+the OUTPUT parameter to accommodate the output stream::
+
+  r = irods.rule.Rule( session, rule_file = 'native1.r',
+             instance_name = 'irods_rule_engine_plugin-irods_rule_language-instance',
+             params={'*stream':'"stdout"'} , output = 'ruleExecOut' )
+  output = r.execute( )
+  if output and len(output.MsParam_PI):
+      buf = output.MsParam_PI[0].inOutStruct.stdoutBuf.buf
+      if buf: print(buf.rstrip(b'\0').decode('utf8'))
+
+(Changing the input value to be squared in this example is left as an exercise for the reader!)
+
+To deal with errors resulting from rule execution failure, two approaches can be taken. Suppose we
+have defined this in the :code:`/etc/irods/core.re` rule-base::
+
+  rule_that_fails_with_error_code(*x) {
+    *y = (if (*x!="") then int(*x) else 0)
+  # if (SOME_PROCEDURE_GOES_WRONG) {
+      if (*y < 0) { failmsg(*y,"-- my error message --"); }  #-> throws an error code of int(*x) in REPF
+      else { fail(); }                                       #-> throws FAIL_ACTION_ENCOUNTERED_ERR in REPF
+  # }
+  }
+
+We can run the rule thus:
+
+>>> Rule( session, body='rule_that_fails_with_error_code(""), instance_name = 'irods_rule_engine_plugin-irods_rule_language-instance',
+        ).execute( r_error = (r_errs:= irods.message.RErrorStack()) )
+
+Where we've used the Python 3.8 "walrus operator" for brevity.  The error will automatically be caught and translated to a
+returned-error stack::
+
+   >>> pprint.pprint([vars(r) for r in r_errs])
+   [{'raw_msg_': 'DEBUG: fail action encountered\n'
+                 'line 14, col 15, rule base core\n'
+                 '        else { fail(); }\n'
+                 '               ^\n'
+                 '\n',
+     'status_': -1220000}]
+
+Note, if a stringized negative integer is given , ie. as a special fail code to be thrown within the rule,
+we must add this code into a special parameter to have this automatically caught as well:
+
+>>> Rule( session, body='rule_that_fails_with_error_code("-2")',instance_name = 'irods_rule_engine_plugin-irods_rule_language-instance'
+...     ).execute( r_error = (r_errs := irods.message.RErrorStack()),
+...           I    acceptable_errors = ( FAIL_ACTION_ENCOUNTERED_ERR, -2))
+
+Because rule is written to emit a custom error message via failmsg int this case, the resulting r_error stack will now include that
+custom error message as a substring::
+
+  >>> pprint.pprint([vars(r) for r in r_errs])
+  [{'raw_msg_': 'DEBUG: -- my error message --\n'
+              'line 21, col 20, rule base core\n'
+              '      if (*y < 0) { failmsg(*y,"-- my error message --"); }  '
+              '#-> throws an error code of int(*x) in REPF\n'
+              '                    ^\n'
+              '\n',
+    'status_': -1220000}]
```

Comment:
four spaces instead of three?

Reply:
W.r.t. the indentation of the block as a whole?  I  think RST allows any degree of indentation to stand for a code example, but I can increase all blocks' indentations if that's better.

Reply:
yes, the left edge.   consistency makes me happy.

i notice the one above (L543) isn't indented at all...

Reply:
L543 is a different scheme not involving indentation... uses '>>>' and '...' to show a Python interactive session. Whereas the indented blocks elsewhere are code examples wherein the indentation shows where the code stops....
Tried indenting L543, ended up with a grey bar to the left which made it look inconsistent on the webpage.  

Yeah.  RST is kind of painful.

Reply:
got it - okay - if it's rendering the way you want it to - let's go with that.